core/src/shred.rs

```diff
@@ -381,10 +427,229 @@ impl Shredder {
-        let final_shred = self.make_final_data_shred();
+        let mut final_shred = self.make_final_data_shred();
+        final_shred.header.last_in_slot = 1;
         self.finalize_data_shred(Shred::LastInSlot(final_shred));
         self.generate_coding_shreds();
     }
+
+    fn fill_in_missing_shreds(
+        shred: &Shred,
+        num_data: usize,
+        num_coding: usize,
+        slot: u64,
+        first_index: usize,
+        expected_index: usize,
+        present: &mut [bool],
+    ) -> (Vec<Vec<u8>>, bool, usize) {
+        let (index, mut first_shred_in_slot) = Self::get_shred_index(shred, num_data);
+
+        let mut missing_blocks: Vec<Vec<u8>> = (expected_index..index)
+            .map(|missing| {
+                present[missing] = false;
+                // If index 0 shred is missing, then first shred in slot will also be recovered
+                first_shred_in_slot |= missing == 0;
+                Shredder::new_empty_missing_shred(
+                    num_data,
+                    num_coding,
+                    slot,
+                    first_index,
+                    index,
+                    missing,
+                )
+            })
+            .collect();
+        let shred_buf = bincode::serialize(shred).unwrap();
+        missing_blocks.push(shred_buf);
+        (missing_blocks, first_shred_in_slot, index)
+    }
+
+    fn new_empty_missing_shred(
+        num_data: usize,
+        num_coding: usize,
+        slot: u64,
+        first_index: usize,
+        index: usize,
+        missing: usize,
+    ) -> Vec<u8> {
+        let missing_shred = if missing == 0 {
+            let mut data_shred = FirstDataShred::default();
+            data_shred.header.data_header.common_header.slot = slot;
+            data_shred.header.data_header.common_header.index = 0;
+            Shred::FirstInSlot(data_shred)
+        } else if missing < first_index + num_data {
+            let mut data_shred = DataShred::default();
+            data_shred.header.common_header.slot = slot;
+            data_shred.header.common_header.index = missing as u32;
+            if index == first_index + num_data - 1 {
+                Shred::LastInFECSet(data_shred)
+            } else {
+                Shred::Data(data_shred)
+            }
+        } else {
+            Shred::Coding(Self::new_coding_shred(
+                slot,
+                missing.saturating_sub(num_data) as u32,
+                num_data,
+                num_coding,
+                missing - first_index - num_data,
+            ))
+        };
+        bincode::serialize(&missing_shred).unwrap()
+    }
+
+    /// Combines all shreds to recreate the original buffer
+    /// If the shreds include coding shreds, and if not all shreds are present, it tries
+    /// to reconstruct missing shreds using erasure
+    /// Note: The shreds are expected to be sorted
+    /// (lower to higher index, and data shreds before coding shreds)
```

Comment:
is that the easiest signature for your customer (blocktree)?

there's an assumption that all the shreds are in the same set?

Reply:
Yes, the assumption is that all shreds belong to the same set. The caller needs to perform additional work.. It sucks a bit. But otherwise it'll add lot more complexity to `deshred()`

Reply:
ok